proposals/kroxylicious-operator-api-alternative.md

```diff
@@ -0,0 +1,654 @@
+
+### VirtualCluster
+
+The VirtualCluster CR is the responsibilty of the the Developer.
+
+Declares a virtualcluster.
+
+A Virtualcluster is associated with exactly one Proxy.
+
+It enumerate the ingresses that are to be used by the virtualcluster.  It also supplies the virtual cluster ingress specific information
+such as the TLS certificates.
+
+The virtualcluster has a reference to a single target cluster which may be expressed using either a reference to a Strimzi Kafka object, or generic bootstraping information.
+
+```yaml
+apiVersion: proxy.kroxylicious.io/v1
+kind: VirtualCluster
+metadata:
+  name: mycluster
+spec:
+  proxyRef:
+    kind: Proxy  # if present must be Proxy, otherwise defaulted
+    group: proxy.kroxylicious.io # if present must be proxy.kroxylicious.io, otherwise defaulted
+    name: myproxy
+ 
+  # list of ingresses the virtual cluster wishes to use.  Each ingress selected must be one associated with
+  # the proxy.  It is legal for the virtualcluster to choose only a subset.
+  ingresses:
+  - name: myclusterip
+    tls:
+      # server certificate - grab bag matched by hostname (what about key/keystore password?)
+      certificateRefs:
+      - kind: Secret # if present must be Secret, otherwise defaulted to Secret
+        group: ""  # if present must be "", otherwise defaulted to ""
+        name: servercert
+        namespace: # namespace of the secret, if omitted assumes namespace of this resource
+     # peer trust
+     trustAnchorRefs:
+      - kind: ConfigMap # if present must be ConfigMap, otherwise defaulted to ConfigMap
+        group: ""  # if present must be "", otherwise defaulted to ""
+        name: servercert
+        namespace: # namespace of the configmap, if omitted assumes namespace of this resource
+      trustOptions:
+       clientAuth: REQUIRED
+    infrastructure:
+      # Controls that let the Developer influence the annotations/labels created on the ingress objects (Service, OpenShift Routes, TLSRoutes etc).
+      # If an annotation or label supplied here duplicates one provided by the ProxyIngress, the one defined by the ProxyIngress takes precendence. 
+      # The specific use case for this is https://docs.openshift.com/container-platform/4.17/security/certificates/service-serving-certificate.html#add-service-certificate_service-serving-certificate
+      # where the developer wishes their VC to use Certificates created by OpenShift (useful for ClusterIP type TLS).
+      annotations:
+         x: y
+      labels:
+         x: y
+
+  # Points to the cluster being proxied.  Can either be Strimzi Kafka resource or endpoint details.
+  targetCluster:
+    # one of: resourceRef or resource
+    resourceRef:
+      kind: Kafka  # must be Kafka
+      group: strimzi.io # must be strimzi.io
+      name: my-cluster
+      listenerName: listener # name of strimzi listener 
+    resource:
+       bootstrap: bootstrap:9092
+       protocol: TCP|TLS
+       nodeIdRanges:
+      - name: mybrokers
```

Comment:
Is there a reason we need to name this?

Reply:
I think we borrowed the idea from the Strimzi model when RangeAwarePortPerNodeCluster scheme was implemented. 

Naming the range isn't required for the function.  We could remove it.